presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1411,48 +1414,84 @@ else if (expression instanceof LongLiteral) {
         return orderByExpressions;
     }
 
-    private List<FieldOrExpression> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
+    private List<List<FieldOrExpression>> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
     {
-        ImmutableList.Builder<FieldOrExpression> groupByExpressionsBuilder = ImmutableList.builder();
-        if (!node.getGroupBy().isEmpty()) {
-            // Translate group by expressions that reference ordinals
-            for (Expression expression : node.getGroupBy()) {
-                // first, see if this is an ordinal
-                FieldOrExpression groupByExpression;
+        List<List<FieldOrExpression>> analyzedGroupingSets;
+        if (node.getGroupBy().isEmpty()) {
+            analyzedGroupingSets = extractAggregates(node).isEmpty() ? ImmutableList.of() : ImmutableList.of(ImmutableList.of());
+            analysis.setGroupingSets(node, analyzedGroupingSets);
+            return analyzedGroupingSets;
+        }
 
-                if (expression instanceof LongLiteral) {
-                    long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > outputExpressions.size()) {
-                        throw new SemanticException(INVALID_ORDINAL, expression, "GROUP BY position %s is not in select list", ordinal);
-                    }
+        List<Set<List<Expression>>> enumeratedGroupingSets = node.getGroupBy().stream()
+                .map(GroupingElement::enumerateGroupingSets)
+                .distinct()
```

Comment:
Do we need to canonicalize the column names? Otherwise, (a,b) will be considered different from (b,a)